File: content/5-concurrency/2-timeout-cancel/go/main.go
package main

import (
	"context"
	"fmt"
	"time"
)

func slowTask(ctx context.Context) (string, error) {
	select {
	case <-time.After(3 * time.Second):
		return "finished", nil
	case <-ctx.Done():
		return "", ctx.Err()
	}
}

func main() {
	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
	defer cancel()

	result, err := slowTask(ctx)
	if err != nil {
		fmt.Println("timeout:", err)
		return
	}
	fmt.Println(result)
}

File: content/5-concurrency/2-timeout-cancel/python/main.py
import asyncio


async def slow_task() -> str:
    await asyncio.sleep(3)
    return "finished"


async def main() -> None:
    try:
        result = await asyncio.wait_for(slow_task(), timeout=1.0)
        print(result)
    except asyncio.TimeoutError:
        print("timeout: TimeoutError")


asyncio.run(main())

